rfd/0091-session-streaming.md

```diff
@@ -0,0 +1,238 @@
+
+```
+func StreamSessionEvents(
+    ctx context.Context, 
+    sessionID session.ID, 
+    startIndex int64,
+  ) (chan apievents.AuditEvent, chan error) {
+```
+
+A gRPC wrapper is also available:
+
+```
+  rpc StreamSessionEvents(StreamSessionEventsRequest) returns (stream events.OneOf);
+```
+
+Given a session ID and a start index, a channel of `AuditEvent`s is returned.
+Consumers can control the speed of playback by limiting how fast they consume
+from this channel. Each event has a timestamp indicating the number of 
+milliseconds that have elapsed since the beginning of the session, so consumers
+can implement "real time" playback by sleeping for the correct amount of time
+before consuming the next event. Speeding up playback can be accomplished by 
+sleeping for less time than required.
+
+Skipping forward in a session can be done efficiently by playing back events as
+fast as possible with no timing delay in between events until the desired
+position is reached.
+
+Skipping backward in a session is the most difficult function to  implement, as
+this API offers no way to go back. Once an event is consumed from 
+the channel it is gone. While it is possible to keep some sort of buffer for the
+last N events, this approach is not suitable here for several reasons:
+
+- There is not an obvious buffer size to use. Select too small of a buffer, and
+  you end up needing to start from the beginning anyway. Select too large of a
+  buffer and you end up in the same state we are in today 
+  (too much memory usage, high latency, etc)
+- The state of the terminal at any point in time is a function of all of the
+  events that have occurred up until that time, so you always need to play from
+  the beginning to accurately depict the session. This is true for both TTY 
+  sessions and desktop sessions.
+
+Here we propose the naive approach of rewinding by:
+
+1. restarting the stream from the beginning
+2. "fast forwarding" up to the desired rewind point by playing back events 
+   without inserting a timing delay
+3. resuming playback from this point at regular speed
+
+### `tsh`
+
+For `tsh` we can leverage the gRPC call that wraps `StreamSessionEvents`.
+`tsh` will perform playback in a background goroutine, and listen for keypresses
+to control playback (play, pause, skip forward, etc).
+
+`tsh` supports the following playback commands:
+
+- `ctrl+c` / `ctrl+d`: stop playing
+- `space`: play/pause
+- `left` / `down`: rewind
+- `right` / `up`: fast-forward
+
+This will be a relatively isolated change, as the `tsh` player code already
+ranges over a slice of events. The loop will need to range over a channel of
+events instead.
+
+### Web UI
+
+Today, the web UI performs an API call to 
+`/webapi/sites/:site/sessions/:sid/stream`. Unlike other APIs, this endpoint
+does not return JSON - it returns the text contents directly. This API call
+is synchronous and does not support streaming, it writes the entire session
+in a single response.
```

Comment:
it doesn't matter much, just more fyi,  i think `writes entire session in a single response` is slightly inaccurate, it returns chunks based on offset and bytes? [e.g. how frontend fetches](https://github.com/gravitational/webapps/blob/4da005d774cff0820b50439541a3d6ee1f89d0b3/packages/teleport/src/lib/term/ttyPlayerEventProvider.js#L79)

Reply:
Noted, thanks!